# 11 - Counterfactual Calculation

In [38]:
import pandas as pd
import pickle

In [39]:
# Display all rows and columns
pd.set_option('display.max_colwidth', None)  # Show full content in each cell
pd.set_option('display.max_rows', None)      # Show all rows
pd.set_option('display.max_columns', None)   # Show all columns

In [40]:
df = pd.read_parquet("../data/processed/X_test.parquet")

In [41]:
df.head()

Position_IT Support  Position_Production Technician I  \
654                     0                                 1   
79                      0                                 1   
2775                    0                                 0   
1785                    0                                 0   
663                     0                                 1   

      Position_Area Sales Manager  Position_Production Manager  \
654                             0                            0   
79                              0                            0   
2775                            0                            0   
1785                            0                            0   
663                             0                            0   

      Position_Production Technician II  Position_Sales Manager  \
654                                   0                       0   
79                                    0                       0   
2775                                  1                       0   
1785                                  0                       0   
663                                   0                       0   

      Position_Enterprise Architect  Position_Network Engineer  \
654                               0                          0   
79                                0                          0   
2775                              0                          0   
1785                              0                          0   
663                               0                          0   

      Position_Sr. Network Engineer  Position_Database Administrator  \
654                               0                                0   
79                                0                                0   
2775                              0                                0   
1785                              0                                0   
663                               0                                0   

      Position_Data Analyst  Position_Software Engineer  Position_Sr. DBA  \
654                       0                           0                 0   
79                        0                           0                 0   
2775                      0                           0                 0   
1785                      0                           0                 0   
663                       0                           0                 0   

      Position_Sr. Accountant  Position_Administrative Assistant  \
654                         0                                  0   
79                          0                                  0   
2775                        0                                  0   
1785                        0                                  0   
663                         0                                  0   

      Position_Accountant I  Position_Shared Services Manager  \
654                       0                                 0   
79                        0                                 0   
2775                      0                                 0   
1785                      0                                 0   
663                       0                                 0   

      Position_IT Director  Position_CIO  Position_Principal Data Architect  \
654                      0             0                                  0   
79                       0             0                                  0   
2775                     0             0                                  0   
1785                     0             0                                  0   
663                      0             0                                  0   

      Position_IT Manager - DB  Position_IT Manager - Support  \
654                          0                              0   
79                           0                              0   
2775                         0                              0   

In [42]:
# Load the model
model_path = "../models/xgb_model.pkl"
with open(model_path, 'rb') as file:
    xgb_model = pickle.load(file)

In [43]:
probs = xgb_model.predict_proba(df)[:, 1]  # Get the probability of the positive class

In [44]:
# Append probabilities as a new column in the dataframe
df['probability'] = probs

In [45]:
# Define thresholds for filtering
threshold_lower = 0.45
threshold_upper = 0.55

In [46]:
# Filter rows based on probabilities being close to the threshold range (e.g., 0.45 to 0.5)
filtered_df = df[(df['probability'] >= threshold_lower) & (df['probability'] <= threshold_upper)]

In [62]:
df.YearsExperience.value_counts()

YearsExperience
13    77
10    70
11    51
12    40
14    24
9     18
8     14
16     8
7      4
15     3
17     3
19     1
Name: count, dtype: int64

In [47]:
filtered_df.head()

Position_IT Support  Position_Production Technician I  \
2861                    0                                 1   
2951                    0                                 0   
2867                    0                                 1   
2873                    0                                 1   
1391                    0                                 0   

      Position_Area Sales Manager  Position_Production Manager  \
2861                            0                            0   
2951                            0                            0   
2867                            0                            0   
2873                            0                            0   
1391                            0                            0   

      Position_Production Technician II  Position_Sales Manager  \
2861                                  0                       0   
2951                                  0                       0   
2867                                  0                       0   
2873                                  0                       0   
1391                                  0                       0   

      Position_Enterprise Architect  Position_Network Engineer  \
2861                              0                          0   
2951                              0                          0   
2867                              0                          0   
2873                              0                          0   
1391                              0                          0   

      Position_Sr. Network Engineer  Position_Database Administrator  \
2861                              0                                0   
2951                              0                                0   
2867                              0                                0   
2873                              0                                0   
1391                              0                                0   

      Position_Data Analyst  Position_Software Engineer  Position_Sr. DBA  \
2861                      0                           0                 0   
2951                      1                           0                 0   
2867                      0                           0                 0   
2873                      0                           0                 0   
1391                      0                           0                 0   

      Position_Sr. Accountant  Position_Administrative Assistant  \
2861                        0                                  0   
2951                        0                                  0   
2867                        0                                  0   
2873                        0                                  0   
1391                        0                                  0   

      Position_Accountant I  Position_Shared Services Manager  \
2861                      0                                 0   
2951                      0                                 0   
2867                      0                                 0   
2873                      0                                 0   
1391                      0                                 0   

      Position_IT Director  Position_CIO  Position_Principal Data Architect  \
2861                     0             0                                  0   
2951                     0             0                                  0   
2867                     0             0                                  0   
2873                     0             0                                  0   
1391                     0             0                                  0   

      Position_IT Manager - DB  Position_IT Manager - Support  \
2861                         0                              0   
2951                         0                              0   
2867                         0                              0   

In [49]:
row_to_modify = filtered_df.iloc[0].copy()

In [50]:
# Display the original row and its probability
print("Original Row:")
print(row_to_modify)

Original Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000000
Position_Sr. Accountant              

In [51]:
# Modify the row: change RaceDesc_White from 1 to 0 and RaceDesc_Asian from 0 to 1
if 'RaceDesc_White' in row_to_modify.index and 'RaceDesc_Asian' in row_to_modify.index:
    row_to_modify['RaceDesc_White'] = 0
    row_to_modify['RaceDesc_Asian'] = 1

In [52]:
# Display the modified row
print("\nModified Row:")
print(row_to_modify)


Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000000
Position_Sr. Accountant             

In [53]:
# Convert the row back into a DataFrame (model expects a DataFrame for predictions)
modified_row_df = row_to_modify.to_frame().T

In [55]:
probability = modified_row_df.probability

In [59]:
probability

2861    0.488037
Name: probability, dtype: float64

In [56]:
modified_row_df.drop(columns="probability", inplace=True)

In [57]:
# Recalculate the probability for the modified row
counterfactual_probability = xgb_model.predict_proba(modified_row_df)[:, 1][0]

In [58]:
# Display the counterfactual probability
print("\nCounterfactual Probability:")
print(counterfactual_probability)


Counterfactual Probability:
0.99927


The counterfactual probability of **0.99927** shows that changing `RaceDesc_White` to 0 and `RaceDesc_Asian` to 1 caused the model to predict the positive class with almost 100% confidence. This suggests the model might be overly sensitive to these features, possibly due to their high importance or correlations in the training data.

Modifiable Attributes : gender, age, race

In [66]:
def modify_and_recalculate_probability(row: pd.Series, 
                                      xgb_model: object, 
                                      attribute: str, 
                                      new_value: int | str) -> dict:
    """
    Modify a single attribute of a selected row, recalculate probability, and return results.

    Parameters:
    row (pd.Series): The original row selected by the user.
    xgb_model: The trained XGBoost model for predicting probabilities.
    attribute (str): The attribute to modify ('Sex', 'Race', or 'YearsExperience').
    new_value: The new value to set for the selected attribute.
        - For 'Sex': 0 (Female) or 1 (Male)
        - For 'Race': 'White', 'Black', or 'Asian'
        - For 'YearsExperience': 7, 13, or 17

    Returns:
    dict: A dictionary containing the original probability, modified row, and counterfactual probability.
    """
    import pandas as pd

    # Validate inputs
    if attribute not in ['Sex', 'Race', 'YearsExperience']:
        raise ValueError("Invalid attribute. Must be 'Sex', 'Race', or 'YearsExperience'.")

    if attribute == 'Sex' and new_value not in [0, 1]:
        raise ValueError("Invalid value for 'Sex'. Must be 0 (Female) or 1 (Male).")

    if attribute == 'Race' and new_value not in ['White', 'Black', 'Asian']:
        raise ValueError("Invalid value for 'Race'. Must be 'White', 'Black', or 'Asian'.")

    if attribute == 'YearsExperience' and new_value not in [7, 13, 17]:
        raise ValueError("Invalid value for 'YearsExperience'. Must be 7, 13, or 17.")

    # Copy the row to avoid modifying the original data
    modified_row = row.copy()

    # Save the original probability
    original_probability = modified_row.get('probability', None)
    if original_probability is None:
        raise KeyError("The column 'probability' is missing in the row.")

    # Modify the specified attribute
    if attribute == 'Sex':
        modified_row['Sex'] = new_value
    elif attribute == 'Race':
        # Reset all race columns to 0 and set the selected race to 1
        race_columns = ['RaceDesc_White', 'RaceDesc_Black or African American', 'RaceDesc_Asian']
        for col in race_columns:
            if col in modified_row:
                modified_row[col] = 0

        race_column_mapping = {
            'White': 'RaceDesc_White',
            'Black': 'RaceDesc_Black or African American',
            'Asian': 'RaceDesc_Asian'
        }
        race_col = race_column_mapping.get(new_value)
        if race_col in modified_row:
            modified_row[race_col] = 1
    elif attribute == 'YearsExperience':
        modified_row['YearsExperience'] = new_value

    # Convert the modified row to a DataFrame (XGBoost model expects this format)
    modified_row_df = modified_row.to_frame().T
    modified_row_df.drop(columns="probability", inplace=True)

    # Recalculate probability for the modified row
    counterfactual_probability = xgb_model.predict_proba(modified_row_df)[:, 1][0]

    return {
        "original_probability": original_probability,
        "counterfactual_probability": counterfactual_probability,
        "modified_row": modified_row
    }

In [67]:
result = modify_and_recalculate_probability(row_to_modify, xgb_model, 'Race', 'Asian')
print(result)

{'original_probability': np.float64(0.4880373179912567), 'counterfactual_probability': np.float32(0.99927), 'modified_row': Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000


In [68]:
# Filter for a row where 'RaceDesc_White' is 1
white_row = df[df['RaceDesc_White'] == 1].iloc[0]

# Change race from 'White' to 'Black'
result = modify_and_recalculate_probability(white_row, xgb_model, 'Race', 'Black')

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])

Original Probability: 0.6975823640823364
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               1.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

In [71]:
# Filter for a row where 'Sex' is 1 (Male)
male_row = filtered_df[filtered_df['Sex'] == 1].iloc[0]

# Change gender from Male to Female
result = modify_and_recalculate_probability(male_row, xgb_model, 'Sex', 0)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])

Original Probability: 0.46454301476478577
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           1.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.00

In [82]:
# Filter for a row with a specific value for 'YearsExperience' (e.g., 7, 13, or 17)
years_row = df[df['YearsExperience'] == 7].iloc[1]

# Change 'YearsExperience' to a new value (e.g., 13)
result = modify_and_recalculate_probability(years_row, xgb_model, 'YearsExperience', 13)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])


Original Probability: 0.7582640051841736
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                0.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

In [83]:
# Filter for a row with a specific value for 'YearsExperience' (e.g., 7, 13, or 17)
years_row = filtered_df[filtered_df['YearsExperience'] == 13].iloc[0]

# Change 'YearsExperience' to a new value (e.g., 17)
result = modify_and_recalculate_probability(years_row, xgb_model, 'YearsExperience', 17)

print("Original Probability:", result['original_probability'])
print("Modified Row:")
print(result['modified_row'])
print("Counterfactual Probability:", result['counterfactual_probability'])


Original Probability: 0.4880373179912567
Modified Row:
Position_IT Support                                             0.000000
Position_Production Technician I                                1.000000
Position_Area Sales Manager                                     0.000000
Position_Production Manager                                     0.000000
Position_Production Technician II                               0.000000
Position_Sales Manager                                          0.000000
Position_Enterprise Architect                                   0.000000
Position_Network Engineer                                       0.000000
Position_Sr. Network Engineer                                   0.000000
Position_Database Administrator                                 0.000000
Position_Data Analyst                                           0.000000
Position_Software Engineer                                      0.000000
Position_Sr. DBA                                                0.000

By changing individual attributes like race, gender, or years of experience, we observed significant shifts in the model's predictions. These results highlight the sensitivity of the model to specific features and emphasize the need to evaluate feature importance and potential biases. This demonstrates the impact of single-value modifications on prediction outcomes.